### PR TITLE
Add some bin/ helpers

### PR DESCRIPTION
**Pull request by elia**
**Description**

Add a bunch of bin files that anyone can expect to be present when working on a gem and that can make the life a bit easier for the Solidus developer. 

- **bin/setup** will do the usual gem setup, with bundler, it tries with `install` first, then falls back to `update`
- **bin/console** will just start irb with `bundler` and `solidus` already required
- **bin/rake** it's a simple binstup, for parity with the latest habits of Rails
- **bin/sandbox** formerly `rake sandbox` will (re)create the sandbox app, any argument will be passed to the `spree:install` generator
- **bin/rails** will forward everything to the `bin/rails` of the sandbox, will also create the sandbox if it's missing
- **bin/rspec** will run specs from the root, changing directory as appropriate for each group of specs (see the commit description for more details).

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [x] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)
- [x] I have attached screenshots to this PR for visual changes (if needed)

### CODE REVIEW

File: bin/sandbox [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  --user_class=Spree::User \
+  --enforce_available_locales=true \
+  $@
+
+unbundled bin/rails generate solidus:auth:install
+
+echo "
+🚀 This app is intended for test purposes. If you're interested in running
+🚀 Solidus in production, visit:
+🚀 https://guides.solidus.io/developers/getting-started/first-time-installation"
```
Comment (tvdeyen) [inline bin/sandbox]:
```
› bin/sandbox
Traceback (most recent call last):
-e:1:in `<main>': undefined method `with_unbundled_env' for Bundler:Module (NoMethodError)
Did you mean?  with_clean_env
```

File: bin/rails [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    warn "Automatic creation of the sandbox app failed"
+    exit 1
+  end
+end
+
+puts "Running Rails from: #{app_root}"
+Dir.chdir app_root
+
+exec "#{__dir__}/unbundled", 'bin/rails', *ARGV
+
```
Comment (tvdeyen) [inline bin/rails]:
```
› bin/rails                       
Traceback (most recent call last):
-e:1:in `<main>': undefined method `with_unbundled_env' for Bundler:Module (NoMethodError)
Did you mean?  with_clean_env
Automatic creation of the sandbox app failed
```

**Comment by elia** [comment]
@tvdeyen thanks for the review, the problem you're facing is due to a deprecation within bundler that I tried to fix 

https://github.com/rubygems/bundler/blob/bada03dd6d4d15828fb5b2cf7f744948e88a69a3/lib/bundler.rb#L367-L377

I'm ok with either support both methods, go back to the older one, or require bundler 2.1, let me know which of the two makes more sense to you, meanwhile installing bundler 2.1 should allow you to try the helpers out.

**Comment by tvdeyen** [comment]
@elia since this only effects local development I am fine with both as well. Just make sure that it is obvious that we require an updated bundler version for local development. 

Thanks for this, this is really great 👍

**Comment by elia** [comment]
@tvdeyen I've updated the code to install 2.1 or above as part of bin/setup and bin/sandbox, that should make the experience smooth for everyone.

**Comment by kennyadsl** [comment]
Thanks @elia, just one question: what is a use case of `bin/console` exactly?

**Comment by elia** [comment]
> Thanks @elia, just one question: what is a use case of `bin/console` exactly?

@kennyadsl that's a good point, I initially included it because it comes with the standard `bundle gem` template expecting others to find it useful for something 😅

Anyway, after further inspection, it was clear that had no real use-case… so I have included a helper that allows the initialization of the dummy app from within the console:
- it relies on DATABASE_URL but defaults to an in-memory sqlite
- it can be easily initialized and migrated
- it features inline instructions
- it allows interaction with all of solidus classes (after initialization) and models (after db migration)

File: bin/console [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+      class << (Rails.application)
+        # Allow to run migrations from the console
+        def migrate!
+          ActiveRecord::Migrator.migrations_paths =
+            Rails.application.migration_railties.flat_map do |engine|
+              (engine.paths['db/migrate'] if engine.respond_to?(:paths)).to_a
+            end
+          ActiveRecord::Migration.verbose = false
+          ActiveRecord::Base.connection.migration_context.run(:up, '20160101010000_solidus_one_four'.to_i)
```
Comment (kennyadsl) [inline bin/console]: Why this up?

Comment (elia) [reply bin/console]: Without running that initial migration in isolation it fails with this error:

```
SQLite3::SQLException: no such table: spree_users
```

**Comment by kennyadsl** [comment]
@gmacdougall @tvdeyen Since your review we also added support to easily load the DummyApp in the `bin/console` (not `bin/rails console`) task, otherwise it's quite useless. Would you mind re-review the last two commits?

File: bin/rails (hunk 2 above)
Comment (tvdeyen) [reply bin/rails]: Still have this issue

```
› bin/sandbox
Successfully installed bundler-2.1.4
1 gem installed
Traceback (most recent call last):
-e:1:in `<main>': undefined method `with_unbundled_env' for Bundler:Module (NoMethodError)
Did you mean?  with_clean_env
```

File: bin/sandbox (hunk 1 above)
Comment (tvdeyen) [reply bin/sandbox]: Still having this issue:

```
› bin/sandbox
Successfully installed bundler-2.1.4
1 gem installed
Traceback (most recent call last):
-e:1:in `<main>': undefined method `with_unbundled_env' for Bundler:Module (NoMethodError)
Did you mean?  with_clean_env
```

**Comment by elia** [comment]
@tvdeyen is it possible that your main Gemfile.lock is still referencing v1? Maybe we should add `bundle update --bundler` to `bin/setup`…

**Comment by elia** [comment]
@tvdeyen I think it's fixed now, `bin/sandbox` will check bundler's version and redirect the user to run `bin/setup` to fix the problem. `bin/setup` will install bundler 2.1 and update the lock file if bundler is older than 2.1.

```shell
$ bin/sandbox                                                                                                             ~/C/N/solidus
The sandbox requires at least Bundler 2.1, please run bin/setup to update it.

$ bin/setup                                                                                                               ~/C/N/solidus
== Installing dependencies ==
$ gem install bundler -v"~> 2.1" --conservative
Successfully installed bundler-2.1.4
Parsing documentation for bundler-2.1.4
Done installing documentation for bundler after 2 seconds
1 gem installed
$ bundle update --bundler
...
Using bundler 2.1.4
...
Warning: the lockfile is being updated to Bundler 2, after which you will be unable to return to Bundler 1.
Bundle updated!
$ bundle check
The dependency jruby-openssl (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for java. To add those platforms to the bundle, run `bundle lock --add-platform java`.
The dependency activerecord-jdbcsqlite3-adapter (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for java. To add those platforms to the bundle, run `bundle lock --add-platform java`.
The Gemfile's dependencies are satisfied

$ bin/setup                                                                                                               ~/C/N/solidus
== Installing dependencies ==
$ bundle check
The dependency jruby-openssl (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for java. To add those platforms to the bundle, run `bundle lock --add-platform java`.
The dependency activerecord-jdbcsqlite3-adapter (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for java. To add those platforms to the bundle, run `bundle lock --add-platform java`.
The Gemfile's dependencies are satisfied

$ bin/sandbox                                                                                                             ~/C/N/solidus
      create  
      create  README.md
      create  Rakefile
...

**Comment by mabry1985** [comment]
Sorry if this isn't the right place to post this, but I'm trying to get a sandbox setup and am running into a bundler error that @elia worked on and I'm not sure where to go from here. It's been a few months since using Rails regularly so maybe I'm missing something basic, but I've been stuck on this for awhile, so thought I'd reach out for advice. 

```
$ bin/sandbox
The sandbox requires at least Bundler 2.1, please run bin/setup to update it.

$ bin/setup
== Installing dependencies ==
$ gem install bundler -v"~> 2.1" --conservative
$ bundle update --bundler
The dependency jruby-openssl (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for java. To add those platforms to the bundle, run `bundle lock --add-platform java`.
The dependency activerecord-jdbcsqlite3-adapter (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for java. To add those platforms to the bundle, run `bundle lock --add-platform java`.
...
Bundle updated!
$ bundle check
The dependency jruby-openssl (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for java. To add those platforms to the bundle, run `bundle lock --add-platform java`.
The dependency activerecord-jdbcsqlite3-adapter (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for java. To add those platforms to the bundle, run `bundle lock --add-platform java`.
The Gemfile's dependencies are satisfied

$ bundler -v
Bundler version 2.1.4

$ bin/sandbox            
The sandbox requires at least Bundler 2.1, please run bin/setup to update it.
```

**Comment by elia** [comment]
> ```
> gem update --system
> ```
> 
> 
> This is what I had to do in my case. Decided to update in case someone else has the same issue. Remove if unhelpful. Cheers.

@mabry1985 thanks for posting the solution! I guess this will be very helpful for people bumping into the same problem 👍

**Pull request merged**